src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_black_knight.cpp

```diff
@@ -15,337 +15,465 @@
-                bEventInProgress = true;
             }
-        }
-
-        void JustDied(Unit* /*killer*/) override
-        {
-            DoCast(me, SPELL_KILL_CREDIT);
-
-            instance->SetData(BOSS_BLACK_KNIGHT, DONE);
+            DoMeleeAttackIfReady();
         }
     };
 
     CreatureAI* GetAI(Creature* creature) const override
     {
-        return GetTrialOfTheChampionAI<boss_black_knightAI>(creature);
+        return GetTrialOfChampionAI<npc_risen_ghoulAI>(creature);
     }
 };
 
-class npc_risen_ghoul : public CreatureScript
+class boss_black_knight : public CreatureScript
 {
 public:
-    npc_risen_ghoul() : CreatureScript("npc_risen_ghoul") { }
+    boss_black_knight() : CreatureScript("boss_black_knight") { }
 
-    struct npc_risen_ghoulAI : public ScriptedAI
+    struct boss_black_knightAI : public BossAI
     {
-        npc_risen_ghoulAI(Creature* creature) : ScriptedAI(creature)
+        boss_black_knightAI(Creature* creature) : BossAI(creature, DATA_BLACK_KNIGHT)
         {
             Initialize();
+            me->SetFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_NON_ATTACKABLE);
+            me->SetReactState(REACT_PASSIVE);
         }
 
         void Initialize()
         {
-            uiAttackTimer = 3500;
+            uiPhase = PHASE_UNDEAD;
+            achievementCredit = true;
         }
 
-        uint32 uiAttackTimer;
+        uint8 uiPhase;
+        bool achievementCredit;
 
         void Reset() override
         {
+            me->SetDisplayId(me->GetNativeDisplayId());
+            SetEquipmentSlots(false, EQUIP_SWORD, EQUIP_NO_CHANGE, EQUIP_NO_CHANGE);
+            _Reset();
             Initialize();
         }
 
+        void EnterEvadeMode(EvadeReason why = EVADE_REASON_OTHER) override
+        {
+            BossAI::EnterEvadeMode(why);
+            me->RemoveFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_NON_ATTACKABLE | UNIT_FLAG_IMMUNE_TO_PC);
+        }
+
+        // Events are not resetted in this function
+        void LoadEvents()
+        {
+            if (uiPhase == PHASE_GHOST)
+            {
+                // Phase 3
+                events.ScheduleEvent(EVENT_DEATH_BITE, 5000);
+                events.ScheduleEvent(EVENT_MARKED_DEATH, 20000);
+            }
+            else
+            {
+                // Phase 1 and 2
+                events.ScheduleEvent(EVENT_ICY_TOUCH, 6000);
+                if (uiPhase == PHASE_UNDEAD)
+                {
+                    // Phase 1 only
+                    events.ScheduleEvent(EVENT_DEATH_RESPITE, 8000);
+
+                    // Raising announcer as a ghoul
+                    if (Creature* announcer = instance->GetCreature(DATA_ANNOUNCER))
+                    {
+                        if (announcer->GetEntry() == NPC_JAEREN)
+                            announcer->CastSpell(me->GetVictim(), SPELL_RAISE_JAEREN, false);
+                        else
+                            announcer->CastSpell(me->GetVictim(), SPELL_RAISE_ARELAS, false);
```

Comment:
I think those spells are meant to be casted by the black knight, and require some conditions (if they are not already there).